### ISSUE TITLE
Does not converge under pytorch 1.0 and the latest dependencies

### ISSUE THREAD

**Issue by auspicious3000** [r9y9/wavenet_vocoder#143]
On and before **Feb 9, 2019**, the pytorch version is already 1.0.0

What I did:

virtualenv -p python3 vocoder
cd vocoder && . bin/activate
git clone https://github.com/r9y9/wavenet_vocoder && cd wavenet_vocoder
pip install -e ".[train]"
python train.py --data-root=vctk16-spe-f0s --preset=presets/vctk16_sptk_mixture.json --checkpoint-dir=spe_ckpts/ 

It is conditioned on spectral envelope and f0s, and it converges to around 6.4 in 20000 steps.

However, on and after **March 11**, I did **EXACTLY** the same thing, but the loss stays around 11.8 forever (80000 steps). 
I repeated the same procedures on 4 different linux machines with multiple gpus, but the behaviors are the same. 
I tried to start training from a checkpoint, but the loss stays around 6.4 forever, it seems the gradients are not propagated back.  
I repeated the same procedures on an experiment conditioned on 80-dim tacotron mel-spectrum that has converged in Oct. 2018, but now it does not converge anymore. The loss is always 11.8 ish.
So, I guess this is probably an issue caused by a recent upgrade in one of the dependencies. 
I tried to downgrade a few dependencies, but none of them worked until I downgrade pytorch to 0.4.1., which is much slower than 1.0.0

**Issue Summary:**
The wavenet_vocoder used to work with pytorch 1.0.0 without any problems if installed and configured on and before Feb 9 , 2019. It converged so well in the past that produced high quality audios. 
Now, after re-installation on March 11, 2019, it stops to work with pytorch 1.0.0 and the latest dependencies. The loss keeps wiggling around 11.8. 

_I asked one of my colleagues to do the experiment himself, and he confirmed that his experiment does not converge either. And the loss also stays around 11.8. He configured wavenet_vocoder using conda instead of virtualenv._

**Comment by r9y9** [comment]
Hi, thank you for raising the issue and detailed investigation. I think this is the same issue as https://github.com/pytorch/pytorch/issues/14960, https://github.com/ksw0306/FloWaveNet/issues/13 and https://github.com/pytorch/pytorch/issues/15716. For now I would recommend you to stick to pytorch 0.4.1, or use single GPU for training. Luckily, the code should run both on pytorch 1.0.0 and 0.4.1 without much effort.

**Comment by r9y9** [comment]
```py
grad_norm = torch.nn.utils.clip_grad_norm_(model.parameters(), hparams.clip_norm)
if np.isnan(grad_norm):
    print("Grad norm is Nan at step {}".format(step))
```

That may help you to check if the grad is actually NaN or not.

**Comment by auspicious3000** [comment]
@r9y9 Thanks for your fast response! 
I have switched to 0.4.1 for now, but it is 50% slower than 1.0.0.
It still puzzles me why it worked well with 1.0.0 before Feb 9.

**Comment by auspicious3000** [comment]
@r9y9  
grad_norm is always 0.0 for all steps.

**However**, I configured the wavenet_vocoder with **pytorch 1.0** in an old python3 virtualenv created last year. That machine has ubuntu16.04 with 4 1080 Ti gpus cards.
This time, it converges under pytorch 1.0. The grad_norm is well above 0.
So, I guess it is not the problem of pytorch 1.0 alone?

**Comment by r9y9** [comment]
Sorry for the late reply. I have added a workaround (https://github.com/r9y9/wavenet_vocoder/commit/6b9c932fdb2e53e406cfe460d700868891a9efb0) to fix multi-GPU training on pytorch >= 1.0.0. I hope this will fix your problem. Please feel free to reopen if the issue persists.

**Comment by 1ytic** [comment]
I managed to train the FloWaveNet in multi-GPU regime using [Apex](https://github.com/NVIDIA/apex) utilities. Probably, it's suitable for wavenet_vocoder too. Here is my PR https://github.com/ksw0306/FloWaveNet/pull/22
